### PR TITLE
Add Deep Residual Network tutorial

### PR DESCRIPTION
**Pull request by mfl28**
This adds the deep residual network tutorial files. I have tried to make everything behave close to how it works in python.

### CODE REVIEW

**Comment by mfl28** [comment]
Ok seems like something was changed in the last libtorch nigthly, and it now complains that c++14 support is needed, see also https://github.com/pytorch/pytorch/issues/30722#issue-532559370.
We could instead download the stable release in the scripts.sh, what do you think?

**Comment by prabhuomkar** [comment]
> We could instead download the stable release in the scripts.sh, what do you think

That's weird. Let's do this as a part of this PR. And we can see if our CI passes.

File: tutorials/intermediate/deep_residual_network/include/cifar10.h [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+
+#include <torch/data/datasets/base.h>
+#include <torch/data/example.h>
+#include <torch/types.h>
+#include <cstddef>
+#include <string>
+
+// Cifar10 dataset
+// based on: https://github.com/pytorch/pytorch/blob/master/torch/csrc/api/include/torch/data/datasets/mnist.h.
+class Cifar10 : public torch::data::datasets::Dataset<Cifar10> {
```
Comment (prabhuomkar) [inline tutorials/intermediate/deep_residual_network/include/cifar10.h]: Rename this to all capitals: `CIFAR10`

File: tutorials/intermediate/deep_residual_network/src/cifar10.cpp [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+        uint32_t image_end = image_start + 3 * kBytesPerChannelPerRow;
+        std::copy(&data_buffer[image_start], &data_buffer[image_end],
+            reinterpret_cast<char*>(images[i].data_ptr()));
+    }
+
+    return {images.to(torch::kFloat32).div_(255), targets.to(torch::kInt64)};
+}
+}  // namespace
+
+Cifar10::Cifar10(const std::string& root, Mode mode) : mode_(mode) {
```
Comment (prabhuomkar) [inline tutorials/intermediate/deep_residual_network/src/cifar10.cpp]: Changes in naming will come here as well and all other places.
Basically do `Cifar10` -> `CIFAR10`

File: tutorials/intermediate/deep_residual_network/src/main.cpp [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+    auto train_dataset = Cifar10(CIFAR_data_path)
+        .map(ConstantPad(4))
+        .map(RandomHorizontalFlip())
+        .map(RandomCrop({32, 32}))
+        .map(torch::data::transforms::Stack<>());
+
+    // Number of samples in the training set
+    auto num_train_samples = train_dataset.size().value();
+
+    auto test_dataset = Cifar10(CIFAR_data_path, Cifar10::Mode::kTest)
```
Comment (prabhuomkar) [inline tutorials/intermediate/deep_residual_network/src/main.cpp]: `Cifar10` -> `CIFAR10`

File: tutorials/intermediate/deep_residual_network/include/cifar10.h [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+ public:
+    // The mode in which the dataset is loaded
+    enum Mode { kTrain, kTest };
+
+    // Loads the Cifar10 dataset from the `root` path.
+    //
+    // The supplied `root` path should contain the *content* of the unzipped
+    // Cifar10 dataset (binary version), available from http://www.cs.toronto.edu/~kriz/cifar.html.
+    explicit Cifar10(const std::string& root, Mode mode = Mode::kTrain);
```
Comment (prabhuomkar) [inline tutorials/intermediate/deep_residual_network/include/cifar10.h]: `Cifar10` -> `CIFAR10`

Comment (mfl28) [reply tutorials/intermediate/deep_residual_network/include/cifar10.h]: Done

File: tutorials/intermediate/deep_residual_network/src/main.cpp (hunk 3 above)
Comment (mfl28) [reply tutorials/intermediate/deep_residual_network/src/main.cpp]: Done

File: tutorials/intermediate/deep_residual_network/src/cifar10.cpp (hunk 2 above)
Comment (mfl28) [reply tutorials/intermediate/deep_residual_network/src/cifar10.cpp]: Done

File: tutorials/intermediate/deep_residual_network/include/cifar10.h (hunk 1 above)
Comment (mfl28) [reply tutorials/intermediate/deep_residual_network/include/cifar10.h]: Done

**Pull request merged**